In [1]:
import pandas as pd

In [2]:
df = pd.read_excel('data/istanbul-dams-daily-occupancy-rates.xlsx')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8409 entries, 0 to 8408
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Tarih         8409 non-null   object 
 1   Ömerli        8409 non-null   float64
 2   Darlık        8409 non-null   float64
 3   Elmalı        8409 non-null   float64
 4   Terkos        8409 non-null   float64
 5   Alibey        8409 non-null   float64
 6   Büyükçekmece  8409 non-null   float64
 7   Sazlıdere     8409 non-null   float64
 8   Kazandere     8409 non-null   float64
 9   Pabuçdere     8409 non-null   float64
 10  Istrancalar   8409 non-null   float64
dtypes: float64(10), object(1)
memory usage: 722.8+ KB


**&emsp;&emsp;! We need to change 'Tarih' column as date type**

<p style="background-image: linear-gradient(to right, #0D9276, #41b675)"> &nbsp </p>

In [3]:
df.head()

Tarih    Ömerli    Darlık    Elmalı    Terkos  Alibey  \
0  2000-10-23 00:00:00  0.652551  0.750151  0.789792  0.708568     0.0   
1  2000-10-24 00:00:00  0.659731  0.773031  0.817187  0.710344     0.0   
2  2000-10-25 00:00:00  0.663327  0.787929  0.826667  0.708568     0.0   
3  2000-10-26 00:00:00  0.662608  0.793250  0.828125  0.706792     0.0   
4  2000-10-27 00:00:00  0.660450  0.795911  0.828125  0.706792     0.0   

   Büyükçekmece  Sazlıdere  Kazandere  Pabuçdere  Istrancalar  
0      0.521442   0.505489   0.057499   0.068553     0.110953  
1      0.521442   0.506952   0.057499   0.069092     0.077226  
2      0.519666   0.506952   0.057499   0.069361     0.105593  
3      0.517889   0.506952   0.057499   0.069624     0.114171  
4      0.516113   0.505489   0.057499   0.066667     0.114017

<p style="background-image: linear-gradient(to right, #0D9276, #41b675)"> &nbsp </p>

In [4]:
pd.DataFrame(df.loc[:,'Ömerli':'Istrancalar'].mean()).T

Ömerli    Darlık    Elmalı    Terkos    Alibey  Büyükçekmece  Sazlıdere  \
0  1.833362  1.597935  1.258037  1.254488  0.887677      1.058058   1.006471   

   Kazandere  Pabuçdere  Istrancalar  
0   0.788519   0.694218     1.122095

**&emsp;&emsp;!Values should be less than one because they represent percentages**

<p style="background-image: linear-gradient(#0D9276, #FFFFFF 10%)"> &nbsp </p>

**&emsp;&emsp;Create mask and see if any of the columns have values above 1**

In [5]:
mask = (df.iloc[:,1:] > 1)    # do not look at date ('Tarih') column
df[mask.any(axis=1)]

Tarih  Ömerli     Darlık     Elmalı     Terkos     Alibey  \
438   2002-01-04 00:00:00    1.00   0.999971   0.572917   0.974614   1.000000   
439   2002-01-05 00:00:00    1.00   0.999706   0.567708   0.968786   0.978092   
440   2002-01-06 00:00:00    1.00   0.995981   0.567708   0.951356   0.957444   
441   2002-01-07 00:00:00    1.00   0.993321   0.567708   0.941698   0.932695   
442   2002-01-08 00:00:00    1.00   0.995449   0.562500   0.945561   0.911390   
...                   ...     ...        ...        ...        ...        ...   
8320             4.8.2023   69.24  52.090000  26.950000  23.100000  17.820000   
8321             5.8.2023   69.09  51.760000  26.410000  22.680000  17.500000   
8322             6.8.2023   68.80  51.350000  25.950000  22.260000  17.340000   
8323             7.8.2023   68.58  51.090000  25.450000  21.840000  17.130000   
8324             8.8.2023   68.36  50.760000  25.290000  21.560000  16.820000   

      Büyükçekmece  Sazlıdere  Kazandere  Pabuçdere  Istrancalar  
438       0.910599       1.00   0.999992   0.334936     0.996586  
439       0.873021       1.00   0.999992   0.333568     0.996586  
440       0.820421       1.00   0.999992   0.306462     0.996586  
441       0.790357       1.00   0.999992   0.325995     0.996586  
442       0.779086       1.00   0.999992   0.325609     0.996586  
...            ...        ...        ...        ...          ...  
8320     16.970000      20.05   7.820000   4.810000    34.470000  
8321     16.670000      19.81   7.770000   4.640000    34.330000  
8322     16.370000      19.62   7.660000   4.260000    34.020000  
8323     16.070000      19.35   7.820000   4.350000    33.710000  
8324     15.770000      19.00   7.940000   4.470000    33.430000  

[1101 rows x 11 columns]

<p style="background-image: linear-gradient(#0D9276, #FFFFFF 10%)"> &nbsp </p>

**&emsp;&emsp;But, we also see the values very close to 1. First only look at the approximate value of 1**

In [6]:
mask = (df.iloc[:,1:] > 1) & (df.iloc[:,1:] < 1.001)
df[mask.any(axis=1)]

Tarih    Ömerli    Darlık    Elmalı    Terkos    Alibey  \
438   2002-01-04 00:00:00  1.000000  0.999971  0.572917  0.974614  1.000000   
439   2002-01-05 00:00:00  1.000000  0.999706  0.567708  0.968786  0.978092   
440   2002-01-06 00:00:00  1.000000  0.995981  0.567708  0.951356  0.957444   
441   2002-01-07 00:00:00  1.000000  0.993321  0.567708  0.941698  0.932695   
442   2002-01-08 00:00:00  1.000000  0.995449  0.562500  0.945561  0.911390   
...                   ...       ...       ...       ...       ...       ...   
3494  2010-05-18 00:00:00  0.885922  0.952348  0.942917  0.953288  0.865214   
3495  2010-05-19 00:00:00  0.883337  0.949155  0.941250  0.953288  0.858987   
3496  2010-05-20 00:00:00  0.880928  0.946495  0.938750  0.953288  0.853818   
3497  2010-05-21 00:00:00  0.878519  0.944367  0.937083  0.953288  0.849688   
3498  2010-05-22 00:00:00  0.876110  0.942238  0.937083  0.951356  0.845576   

      Büyükçekmece  Sazlıdere  Kazandere  Pabuçdere  Istrancalar  
438       0.910599   1.000000   0.999992   0.334936     0.996586  
439       0.873021   1.000000   0.999992   0.333568     0.996586  
440       0.820421   1.000000   0.999992   0.306462     0.996586  
441       0.790357   1.000000   0.999992   0.325995     0.996586  
442       0.779086   1.000000   0.999992   0.325609     0.996586  
...            ...        ...        ...        ...          ...  
3494      0.948282   0.944168   0.982102   1.000000     0.817546  
3495      0.948282   0.941664   0.976400   1.000000     0.790761  
3496      0.946366   0.939326   0.987804   1.000000     0.776677  
3497      0.946366   0.936989   0.996335   1.000000     0.776436  
3498      0.946366   0.939326   0.993898   1.000000     0.785083  

[971 rows x 11 columns]

**&emsp;&emsp;971 rows has values smaller than 1.001 and they should remain as they are**

<p style="background-image: linear-gradient(#0D9276, #FFFFFF 10%)"> &nbsp </p>

**&emsp;&emsp;Update the mask**

In [7]:
mask = df.iloc[:,1:] > 1.001
df[mask.any(axis=1)]

Tarih  Ömerli  Darlık  Elmalı  Terkos  Alibey  Büyükçekmece  \
8195  1.4.2023   60.65   52.05   40.15   32.42   14.48         30.44   
8196  2.4.2023   61.15   52.23   40.26   32.57   14.57         30.44   
8197  3.4.2023   61.50   52.60   40.26   32.72   14.62         30.61   
8198  4.4.2023   61.92   53.04   40.15   32.87   14.82         30.61   
8199  5.4.2023   62.55   53.46   40.15   33.02   14.82         30.94   
...        ...     ...     ...     ...     ...     ...           ...   
8320  4.8.2023   69.24   52.09   26.95   23.10   17.82         16.97   
8321  5.8.2023   69.09   51.76   26.41   22.68   17.50         16.67   
8322  6.8.2023   68.80   51.35   25.95   22.26   17.34         16.37   
8323  7.8.2023   68.58   51.09   25.45   21.84   17.13         16.07   
8324  8.8.2023   68.36   50.76   25.29   21.56   16.82         15.77   

      Sazlıdere  Kazandere  Pabuçdere  Istrancalar  
8195      32.53       9.23       9.35        39.97  
8196      32.47       9.11       8.98        39.39  
8197      32.47       9.48       9.24        38.02  
8198      32.30       9.52       9.18        38.09  
8199      32.36       9.68       9.59        36.66  
...         ...        ...        ...          ...  
8320      20.05       7.82       4.81        34.47  
8321      19.81       7.77       4.64        34.33  
8322      19.62       7.66       4.26        34.02  
8323      19.35       7.82       4.35        33.71  
8324      19.00       7.94       4.47        33.43  

[130 rows x 11 columns]

**&emsp;&emsp;! These values are shown as percentages, these values must be divided by 100 to place them between 0 and 1**

<p style="background-image: linear-gradient(#0D9276, #FFFFFF 10%)"> &nbsp </p>

**&emsp;&emsp;Lines 8195 through 8324 seem to have the same problem**

In [8]:
df.iloc[8190:8200,:]

Tarih   Ömerli   Darlık   Elmalı   Terkos   Alibey  \
8190  2023-03-27 00:00:00   0.5624   0.4731   0.3144   0.3183   0.1165   
8191  2023-03-28 00:00:00   0.5672   0.4749   0.3271   0.3198   0.1196   
8192  2023-03-29 00:00:00   0.5815   0.4911   0.3742   0.3257   0.1310   
8193  2023-03-30 00:00:00   0.5947   0.5087   0.3964   0.3272   0.1409   
8194  2023-03-31 00:00:00   0.6010   0.5146   0.3990   0.3257   0.1428   
8195             1.4.2023  60.6500  52.0500  40.1500  32.4200  14.4800   
8196             2.4.2023  61.1500  52.2300  40.2600  32.5700  14.5700   
8197             3.4.2023  61.5000  52.6000  40.2600  32.7200  14.6200   
8198             4.4.2023  61.9200  53.0400  40.1500  32.8700  14.8200   
8199             5.4.2023  62.5500  53.4600  40.1500  33.0200  14.8200   

      Büyükçekmece  Sazlıdere  Kazandere  Pabuçdere  Istrancalar  
8190        0.2976     0.3241     0.0859     0.0908       0.3961  
8191        0.3010     0.3253     0.0854     0.0895       0.4159  
8192        0.3077     0.3277     0.0920     0.0959       0.4287  
8193        0.3094     0.3270     0.0939     0.0968       0.4251  
8194        0.3061     0.3258     0.0920     0.0952       0.4468  
8195       30.4400    32.5300     9.2300     9.3500      39.9700  
8196       30.4400    32.4700     9.1100     8.9800      39.3900  
8197       30.6100    32.4700     9.4800     9.2400      38.0200  
8198       30.6100    32.3000     9.5200     9.1800      38.0900  
8199       30.9400    32.3600     9.6800     9.5900      36.6600

In [9]:
df.iloc[8320:8330,:]

Tarih     Ömerli     Darlık     Elmalı     Terkos  \
8320             4.8.2023  69.240000  52.090000  26.950000  23.100000   
8321             5.8.2023  69.090000  51.760000  26.410000  22.680000   
8322             6.8.2023  68.800000  51.350000  25.950000  22.260000   
8323             7.8.2023  68.580000  51.090000  25.450000  21.840000   
8324             8.8.2023  68.360000  50.760000  25.290000  21.560000   
8325  2023-08-09 00:00:00   0.680713   0.504690   0.251667   0.211469   
8326  2023-08-10 00:00:00   0.678530   0.501016   0.251042   0.207358   
8327  2023-08-11 00:00:00   0.678530   0.498076   0.252604   0.210083   
8328  2023-08-12 00:00:00   0.679258   0.495872   0.253854   0.205996   
8329  2023-08-13 00:00:00   0.677077   0.493667   0.250417   0.203272   

         Alibey  Büyükçekmece  Sazlıdere  Kazandere  Pabuçdere  Istrancalar  
8320  17.820000     16.970000  20.050000   7.820000   4.810000    34.470000  
8321  17.500000     16.670000  19.810000   7.770000   4.640000    34.330000  
8322  17.340000     16.370000  19.620000   7.660000   4.260000    34.020000  
8323  17.130000     16.070000  19.350000   7.820000   4.350000    33.710000  
8324  16.820000     15.770000  19.000000   7.940000   4.470000    33.430000  
8325   0.165674      0.153209   0.186624   0.080540   0.046380     0.332037  
8326   0.163126      0.150205   0.183873   0.081424   0.045961     0.327096  
8327   0.164655      0.147201   0.183362   0.082667   0.046380     0.305042  
8328   0.163635      0.142696   0.180297   0.084562   0.046798     0.261103  
8329   0.162616      0.139692   0.177821   0.085388   0.046798     0.254997

**&emsp;&emsp;! Date formats are also different**